### PR TITLE
sec(http): JSONP should require trusted resource URLs

### LINKED ISSUE #11328: Inconsistent handling of $sce trustedUrl with $http
```
// Works:
$http.get($sce.trustAsUrl('/')).success(function(d){$log.log(d);});

// Fails:
$http.get($sce.trustAsUrl('/'), {params: {timestamp: 0}}).success(function(d){$log.log(d);});
```

In the first case, when the trusted URL reaches the XmlHttpRequest, its toString method is used to get the wrapped URL. In the second case, the buildUrl function attempts to add parameters, but the given URL is not a string as expected.

I feel that the buildUrl function should reject a non-string argument for url in all cases.

### LINKED ISSUE #11352: $http's JSONP should require trusted resource URLs
$http does not perform any security checks on the URLs that it receives.  It should require URLs that pass checks for `$sce.RESOURCE_URL`.

This should generally not have a huge impact in terms of breaking backwards compatibility.  The cases that break, are where folks are performing JSONP requests to 3rd party servers.  In such cases, it would be a good idea for the apps that do break to configure the $sce whitelist to include that specific trusted 3rd party server to unbreak themselves.

It might be a good idea to backport this to earlier versions if we can feel confident that we won't actually be breaking most applications.

### PR DESCRIPTION
**Pull request by chirayuk**
WORK-IN-PROGRESS.  Do **NOT** merge.  More work needs to be done and the tests are currently broken.
- JSONP should require trusted resource URLs.  This would be a breaking
  change but maybe not too onerous since same origin URLs are trusted in
  the default config and you can easily whitelist any 3rd party URLs you
  trust in one single place (your app/module config.)
- fix a bug where $http can't handle $sce wrapper URLs.

Closes #11352
Closes #11328

### CODE REVIEW

File: test/ng/httpBackendSpec.js [hunk 1]
```diff
@@ -255,7 +256,22 @@ describe('$httpBackend', function() {
   });
 
 
-  describe('JSONP', function() {
+  [true, false].forEach(function(trustAsResourceUrl) {
```
Comment (gkalpak) [inline test/ng/httpBackendSpec.js]: It seems like a good place for the `they()` helper :)

File: test/ng/httpBackendSpec.js [hunk 2]
```diff
@@ -255,7 +256,22 @@ describe('$httpBackend', function() {
...
+    describe('JSONP: trustAsResourceUrl='+trustAsResourceUrl, function() {
+
+
+    function $backendCall() {
+      var args = Array.prototype.slice.call(arguments);
```
Comment (gkalpak) [inline test/ng/httpBackendSpec.js]: There is also the `sliceArgs()` function. (Just saying...)

File: test/ng/httpBackendSpec.js (hunk 1 above)
Comment (gkalpak) [inline test/ng/httpBackendSpec.js]: Or maybe not (since it's on the `describe` level, not the `it` level)...

File: src/ng/httpBackend.js [hunk 3]
```diff
@@ -142,8 +148,8 @@ function createHttpBackend($browser, createXhr, $browserDefer, callbacks, rawDoc
     // - adds and immediately removes script elements from the document
     var script = rawDocument.createElement('script'), callback = null;
     script.type = "text/javascript";
-    script.src = url;
     script.async = true;
+    script.src = url;
```
Comment (caitp) [inline src/ng/httpBackend.js]: phantom edit?

File: src/ng/http.js [hunk 4]
```diff
@@ -1154,7 +1158,7 @@ function $HttpProvider() {
           reqHeaders[(config.xsrfHeaderName || defaults.xsrfHeaderName)] = xsrfValue;
         }
 
-        $httpBackend(config.method, url, reqData, done, reqHeaders, config.timeout,
+        $httpBackend(config.method, (url === $sce.valueOf(origUrl) ? origUrl : url), reqData, done, reqHeaders, config.timeout,
```
Comment (caitp) [inline src/ng/http.js]: is this necessary? why not just `url = $sce.valueOf(origUrl);`

**Comment by caitp** [comment]
are the ng-mock edits enough to not break application tests?

**Comment by caitp** [comment]
@chirayuk it generally looks good ok, a few nits --- I'll leave it up to you whether you think this can land today or not.

**Comment by petebacondarwin** [comment]
@chirayuk - the description says DO NOT MERGE - is this still accurate or are you happy that this is ready to go in now?

**Comment by IgorMinar** [comment]
can we move this to 1.5?

Also, I don't see an easy way to define the whitelist, do we piggy-back on the existing means to define whitelist or is there no current way to whitelist urls that are ok for jsonp?

File: src/ngMock/angular-mocks.js [hunk 5]
```diff
@@ -1095,7 +1095,7 @@ angular.mock.dump = function(object) {
    ```
  */
 angular.mock.$HttpBackendProvider = function() {
-  this.$get = ['$rootScope', '$timeout', createHttpBackendMock];
+  this.$get = ['$sce', '$rootScope', '$timeout', createHttpBackendMock];
```
Comment (petebacondarwin) [inline src/ngMock/angular-mocks.js]: I don't see why the changes in this file are needed...

File: test/ng/httpBackendSpec.js [hunk 6]
```diff
@@ -255,7 +256,22 @@ describe('$httpBackend', function() {
...
+      var url = args[1];
+      if (trustAsResourceUrl || url == null) {
+        args[1] = $sce.trustAsResourceUrl(url);
+        return $backend.apply(null, args);
+      } else {
+        var badUrlPrefix = url ? url.replace('JSON_CALLBACK', '') : '';
+        expect(function() { $backend.apply(null, args); }).toThrowMinErr(
+          '$sce', 'insecurl', 'Blocked loading resource from url not allowed by $sceDelegate policy.  URL: ' + badUrlPrefix);
+      }
+    }
```
Comment (petebacondarwin) [inline test/ng/httpBackendSpec.js]: There is too much work going on here for a test. We should just duplicate what is needed so that we can easily read the tests.

File: src/ng/http.js (hunk 4 above)
Comment (petebacondarwin) [inline src/ng/http.js]: This is related to https://github.com/angular/angular.js/issues/11328

**Comment by petebacondarwin** [comment]
Closing in favour of #15143 which is a rebased and tidied up version of this.
